### PR TITLE
Sp 480 individual comments

### PR DESCRIPTION
**Pull request by Coupey**

### CODE REVIEW

File: app/comments/get.controller.js [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+const { getCommentText } = require('../../common/utils/getCommentText')
+
+const getComments = async (
+  { path, errors, errorSummary, body, renderInfo, params: { planId }, session: { 'x-auth-token': token } },
+  res
+) => {
+  const renderDetails = renderInfo || {}
+
+  renderDetails.nexturl = path.substring(0, path.lastIndexOf('/'))
+  renderDetails.backurl = `${path.substring(0, path.lastIndexOf('/'))}/decisions`
```
Comment (jonhindle) [inline app/comments/get.controller.js]: Duplication of `path.substring(0, path.lastIndexOf('/')`. 
Lots of alternatives, I'd assign a default in the deconstruct and then stick the object together with an assign or the like:
```
const getComments = async (
  { path, errors, errorSummary, body, renderInfo = {}, params: { planId }, session: { 'x-auth-token': token } },
  res
) => {
  const nexturl = path.substring(0, path.lastIndexOf('/'))
  const renderDetails = {...renderInfo, nextUrl, backurl: `${nexturl}/decisions`
```

File: app/comments/get.controller.js [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+  if (body.comments) {
+    renderDetails.comments = body.comments
+  } else {
+    try {
+      const comments = await getSentencePlanComments(planId, token)
+      renderDetails.comments = getCommentText(comments, 'THEIR_SUMMARY')
```
Comment (jonhindle) [inline app/comments/get.controller.js]: 'THEIR_SUMMARY' - I think I'm using this text const's all over the place. I think we create a shared constants file and reference. Gives us a degree of security for any future uses or changes. Worth raising a ticket?

File: app/comments/get.controller.test.js [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+  }
+
+  beforeEach(() => {
+    req.renderInfo = {}
+    delete req.body.comments
+  })
+
+  beforeEach(() => {
+    getSentencePlanComments.mockReset()
+  })
```
Comment (jonhindle) [inline app/comments/get.controller.test.js]: Two before each blocks

File: app/comments/get.controller.test.js [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+    params: {
+      planId: 1,
+    },
+    session: {
+      'x-auth-token': '1234',
+    },
+    body: {},
+    errors: {},
+    errorSummary: {},
+    renderInfo: null,
```
Comment (jonhindle) [inline app/comments/get.controller.test.js]: beforeEach auto re-assigns this to an empty object. We never test a null/undefined scenario; is that important?

File: app/comments/get.controller.test.js [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+      errorSummary: {},
+      comments: false,
+      errors: {},
+    }
+    getSentencePlanComments.mockReturnValueOnce(commentsEmpty)
+    await controller.getComments(req, res)
+    expect(res.render).toHaveBeenCalledWith(`${__dirname}/index`, expected)
+  })
+  it('should set the correct render values when there are existing comments', async () => {
+    delete req.body.comments
```
Comment (jonhindle) [inline app/comments/get.controller.test.js]: I think the before each block (line 30) already handles this.

File: app/comments/post.controller.js [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+    req.renderInfo = { wordsOver: wordsOver > 0 ? wordsOver : 0 }
+    await getComments(req, res)
+  } else {
+    const comment = [
+      {
+        comment: req.body.comments,
+        commentType: 'THEIR_SUMMARY',
+      },
+    ]
+    await setSentencePlanComment(req.params.planid, comment, req.session['x-auth-token'])
```
Comment (jonhindle) [inline app/comments/post.controller.js]: Async functions; wrap in a try catch to cater for error scenarios?

File: app/comments/post.controller.test.js [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+        },
+      ],
+    }
+    expected.body.comments = Array(255).join('word ')
+    expected.tooManyWords = true
+    expected.renderInfo.wordsOver = 4
+    await controller.postComments(req, res)
+    expect(setSentencePlanComment).not.toHaveBeenCalled()
+    expect(getComments).toHaveBeenCalledWith(expected, res)
+  })
```
Comment (jonhindle) [inline app/comments/post.controller.test.js]: Add tests for error Scenario's from async functions.

File: app/decisions/get.controller.js [hunk 8]
```diff
@@ -0,0 +1,28 @@
+const { logger } = require('../../common/logging/logger')
+const { getSentencePlanComments } = require('../../common/data/sentencePlanningApi')
+const { getCommentText } = require('../../common/utils/getCommentText')
+
+const getDecisions = async (
+  { path, errors, errorSummary, body, renderInfo, params: { planId }, session: { 'x-auth-token': token } },
+  res
+) => {
+  const renderDetails = renderInfo || {}
+  let comments = ''
```
Comment (jonhindle) [inline app/decisions/get.controller.js]: This is assigned an array later. Should we assign an empty Array here?

File: app/decisions/get.controller.js [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+  renderDetails.backurl = path.substring(0, path.lastIndexOf('/'))
```
Comment (jonhindle) [inline app/decisions/get.controller.js]: Is there enough commonality to warrant a helper that auto fills back and next urls in the `renderDetails` object?

File: app/decisions/get.controller.test.js [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+  }
+
+  beforeEach(() => {
+    req.renderInfo = {}
+    delete req.body.decisions
+  })
+
+  beforeEach(() => {
+    getSentencePlanComments.mockReset()
+  })
```
Comment (jonhindle) [inline app/decisions/get.controller.test.js]: Two `beforeEach` blocks.

File: app/decisions/get.controller.test.js [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+      errorSummary: {},
+      decisions: false,
+      errors: {},
+    }
+    getSentencePlanComments.mockReturnValueOnce(commentsEmpty)
+    await controller.getDecisions(req, res)
+    expect(res.render).toHaveBeenCalledWith(`${__dirname}/index`, expected)
+  })
+  it('should set the correct render values when there are existing comments', async () => {
+    delete req.body.decisions
```
Comment (jonhindle) [inline app/decisions/get.controller.test.js]: should be covered by the beforeEach block above

File: app/decisions/get.controller.test.js [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+    params: {
+      planid: 1234,
+    },
+    session: {
+      'x-auth-token': '1234',
+    },
+    body: {},
+    errors: {},
+    errorSummary: {},
+    renderInfo: null,
```
Comment (jonhindle) [inline app/decisions/get.controller.test.js]: this gets re-assigned to an object straight away in the `beforeEach`.

File: app/decisions/post.controller.js [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+    req.renderInfo = { wordsOver: wordsOver > 0 ? wordsOver : 0 }
+    await getDecisions(req, res)
+  } else {
+    const comment = [
+      {
+        comment: req.body.decisions,
+        commentType: 'YOUR_SUMMARY',
+      },
+    ]
+    await setSentencePlanComment(req.params.planId, comment, req.session['x-auth-token'])
```
Comment (jonhindle) [inline app/decisions/post.controller.js]: Need wrapping in a try catch to handle error scenario's.

File: app/decisions/post.controller.test.js [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+        },
+      ],
+    }
+    expected.body.decisions = Array(255).join('word ')
+    expected.tooManyWords = true
+    expected.renderInfo.wordsOver = 4
+    await controller.postDecisions(req, res)
+    expect(setSentencePlanComment).not.toHaveBeenCalled()
+    expect(getDecisions).toHaveBeenCalledWith(expected, res)
+  })
```
Comment (jonhindle) [inline app/decisions/post.controller.test.js]: Need to test and handle async call error scenario.

File: app/comments/get.controller.test.js (hunk 4 above)
Comment (Coupey) [reply app/comments/get.controller.test.js]: Could be - I'll add one in

File: app/decisions/get.controller.js (hunk 9 above)
Comment (Coupey) [reply app/decisions/get.controller.js]: Probably is - we've already raised a ticket for determining back/next from a common place so we can address it there

File: app/comments/get.controller.js (hunk 2 above)
Comment (Coupey) [reply app/comments/get.controller.js]: Created as https://dsdmoj.atlassian.net/browse/SP-482

**Pull request merged**